theforeman.org/scripts/debian/stage_packages.sh

```diff
@@ -7,19 +7,41 @@ set -xe
 # * the web node must have the freight class applied
 
 # Find the deps in the build-dir from the previous step
-DEB_PATH=./debian/${os}/build-${project}
+DEB_PATH=$(find ./ -name build-${project})
 
 # Upload all builds to stagingdeb for testing
-echo "scratch build: uploading to stagingdeb/${os}/${repoowner}-${version}"
+HOSTS="web01.osuosl.theforeman.org"
 export RSYNC_RSH="ssh -i /home/jenkins/workspace/staging_key/rsync_freightstage_key"
 USER=freightstage
-HOSTS="web01.osuosl"
-COMPONENT=${repoowner}-${version}
+
+if [[ "$DEB_PATH" == *"plugins"* ]];then
+  # Don't upload scratch builds to deb.tf.o
+  if [ x$repoowner = xtheforeman ] && [ x$pr_number = x ]; then
+    echo "Built from main repo, uploading to deb/plugins/main"
+    export RSYNC_RSH="ssh -i /home/jenkins/workspace/deb_key/rsync_freight_key"
+    USER=freight
+    if [[ $repo == develop ]]; then
+      COMPONENT=nightly
+    else
+      COMPONENT=$repo
+    fi
+  else
+    echo "scratch build: uploading to stagingdeb/plugins/${repoowner}"
+    COMPONENT=${repoowner}
+  fi
+else
+  echo "scratch build: uploading to stagingdeb/${os}/${repoowner}-${version}"
+  COMPONENT=${repoowner}-${version}
+fi
 
 for HOST in $HOSTS; do
   # The path is important, as freight_rsync (which is run on the web node for incoming
   # transfers) will parse the path to figure out the repo to send debs to.
-  TARGET_PATH="${USER}@${HOST}.theforeman.org:rsync_cache/${os}/${COMPONENT}/"
+  if [[ "$DEB_PATH" == *"plugins"* ]];then
+    TARGET_PATH="${USER}@${HOST}:rsync_cache/plugins/${COMPONENT}/"
+  else
+    TARGET_PATH="${USER}@${HOST}:rsync_cache/${os}/${COMPONENT}/"
```

Comment:
For what it's worth, in Debian you typically have a `suite`. This suite is normally the OS release (buster, focal, etc). However, for our plugins we don't differentiate and only have `plugins`. So in scripts you can make this clear by using `$SUITE`. I would do this around line 10 and reuse it around line 17 (`if [[ "$DEB_PATH" == *"plugins"* ]];then`)

Reply:
Do you happen to know *why* we don't differentiate for plugins but do for everything else?

Reply:
To save time. Most plugins aren't really OS specific: they contain gems, but `bundle install` runs on the target host. I think we're counting on plugins not containing any binaries that may be linked to system libraries. If they do, they belong in the dependencies part which is OS specific.

By doing this, we only need a single build for all plugin repos instead of n builds.

Reply:
So in theory, we could save time by ensuring that the only packages in `dependencies` are ones with binary components? For example, could https://github.com/theforeman/foreman-packaging/tree/deb/develop/dependencies/bionic/kafo_parsers be moved to plugins? Could all the hammer plugins?

Reply:
Technically? Yeah. But those don't take that long to build anyways. So your saving won't be much.

Reply:
It would also be confusing if Hammer is in main for RPMs while it's in plugins for Debian. As for kafo_parsers, then you'd have the Foreman repo depend on plugins to build. Essentially the layering is (in order):
* dependencies
* debian
* plugins

When we start building packages for a new distro version, we start with dependencies, then move on to debian and only then plugins. If you need to be aware that you first need to build some packages from plugins, then it becomes harder.